test/unit/controllers/adhoc_payment/submit_amount_controller_test.js

```diff
@@ -0,0 +1,141 @@
+    const priceOverride = 12550
+    before(done => {
+      product = productFixtures.validCreateProductResponse({
+        type: 'ADHOC',
+        product_name: 'Super duper product',
+        service_name: 'Super GOV service',
+        description: 'Super duper product description'
+      }).getPlain()
+      payment = productFixtures.validCreatePaymentResponse({
+        govuk_status: 'SUCCESS',
+        product_external_id: product.external_id,
+        amount: priceOverride
+      }).getPlain()
+      nock(config.PRODUCTS_URL).get(`/v1/api/products/${product.external_id}`).reply(200, product)
+      nock(config.PRODUCTS_URL).post(`/v1/api/products/${product.external_id}/payments`, {price: priceOverride}).reply(200, payment)
+
+      supertest(createAppWithSession(getApp()))
+        .post(paths.adhocPayment.amount.replace(':productExternalId', product.external_id))
+        .send({
+          'payment-amount': '125.50',
+          csrfToken: csrf().create('123')
+        })
+        .end((err, res) => {
+          response = res
+          done(err)
+        })
+    })
+
+    it('should respond with code:303 redirect', () => {
+      expect(response.statusCode).to.equal(303)
+    })
+
+    it('should redirect to the payment card details page', () => {
+      expect(response.headers).to.have.property('location').to.equal(`http://frontend.url/charges/${payment.external_id}`)
+    })
+  })
+
+  describe('when an empty amount is submitted', function () {
+    before(done => {
+      product = productFixtures.validCreateProductResponse({type: 'ADHOC'}).getPlain()
+      nock(config.PRODUCTS_URL).get(`/v1/api/products/${product.external_id}`).reply(200, product)
+      session = getMockSession()
+      supertest(createAppWithSession(getApp(), session))
+        .post(paths.adhocPayment.amount.replace(':productExternalId', product.external_id))
+        .send({
+          csrfToken: csrf().create('123')
+        })
+        .end((err, res) => {
+          response = res
+          done(err)
+        })
+    })
+
+    it('should respond with code:200 OK', () => {
+      expect(response.statusCode).to.equal(200)
+    })
+
+    it('should add a relevant error message to the session \'flash\'', () => {
+      expect(session.flash).to.have.property('genericError')
+      expect(session.flash.genericError.length).to.equal(1)
+      expect(session.flash.genericError[0]).to.equal(`<h2>Use valid characters only</h2> Choose an amount in pounds and pence using digits and a decimal point. For example “10.50”`)
+    })
+  })
+
+  describe('when an invalid amount is submitted', function () {
+    before(done => {
+      product = productFixtures.validCreateProductResponse({type: 'ADHOC'}).getPlain()
+      nock(config.PRODUCTS_URL).get(`/v1/api/products/${product.external_id}`).reply(200, product)
+      session = getMockSession()
+      const invalidAmount = 'GHTR89&&'
+      supertest(createAppWithSession(getApp(), session))
```

Comment:
Using the mock session method does pollute tests globally, and seems to mean that tests don't run individually for some reason. However this is currently the established pattern for testing session cookie contents.

Can you create a story on the backlog for correcting this at some future point?

Reply:
I've created https://payments-platform.atlassian.net/browse/PP-3183